RFC, DO NOT MERGE AS-IS: Streaming shake implementation

nmathewson opened the pull request:
This branch adds support for streaming output from the SHAKE XOF.  For background, see issue #7894.

I have some open questions and to-do items before merging this branch:

1. In order to implement this code, I needed to allow m_sha3.c to use the raw Keccak permutation, as KeccakF1600.  I exposed it from the C implementation and the x86_64 assembly implementation, but my assembly skills are absent:  I don't feel comfortable doing it myself for all the other asm implementations.  Can someone help me out with this?

2. Is DigestFinalXOF() a good name for this function?  It feels strange to have a "Final" function that you can call repeatedly.

3. Should this branch call OPENSSL_cleanse() on the contents of ctx->buf after copying them out, to ensure that they cannot leak, or is that considered unneeded?

4. Should the code take measures to prevent the user from calling DigestUpdate after DigestFinalXOF has been called?

5. I tested this code using a small program, available at https://people.torproject.org/~nickm/volatile/shake_test.c .  Is this the right kind of thing to use as an openssl test, or is there a better place to put the test for this code?

kroeckx commented:
Final at least for me says that no data can be added, so the name
is good for me.

I don't know what our other Init/Update/Final prevent calling
Update after Final. It's of course better to do it, but if the
others don't do this I don't see it as a problem.

nmathewson commented:
> I don't know what our other Init/Update/Final prevent calling Update after Final. It's of course better to do it, but if the others don't do this I don't see it as a problem.

The others do this, but they do it by having EVP_DigestFinal clear the EVP_MD_CTX object. That won't work in this case, since we need to allow EVP_DigestFinalXOF to be called more than once.

paulidale commented:
1. Assembly isn't something I looked deeply enough into.

1. How would this interact with `EVP_DigestFinalXOF` ?  It could be argued that the EVP call is where this should go and the cleanse left for the init or free.  EVP is our official public interface going forward, a non-EVP interface to do something isn't ideal.

1. There should be a cleanse at or before freeing the MD_CTX (and on reinit).  Would a new flag which specifies streaming output be reasonable?  I.e. set this flag on the context and the cleanse and cleanup behaviours are changed from the final to the free (and on init of course).  Just an idea.
 
1. I'd say yes.

1. It could become the basis of a test.  I'd like to see a known answer test case as well.

levitte commented:
1. I can try
2. I assume you mean  `EVP_DigestFinalXOF`...  and seeing how it's otherwise defined, i.e. the pattern we have already established around `Final` functions, I find such a choice horrible.  It's quite possible that we need to think of some other function for digest retrieval in XOF operations...  `EVP_DigestFetch`, perhaps?  Mind you, it still needs that cleanup must be post-poned, and as far as I can see, that may require an explicit call of `EVP_MD_CTX_reset()`.
3. ...
4. Yes
5. There are already dedicated test programs in `test/`, what's one more? :wink:  However, we do run quite a lot of EVP related tests in `test/evp_test.c`, driven with vector data.

nmathewson commented:
@paulidale, @levitte:
re 2: Yes, I did mean EVP_DigestFinalXOF. Some possible alternative names are EVP_DigestExtractXOF, EVP_DigestSqueezeXOF, EVP_DigestFetchXOF. Do any of them sound good?


@paulidale:
re 3: I agree with the cleanse, but I was asking about the case where we have cached intermediate data in a buffer, and we are copying it out to the caller.  It will be overwritten when we cleanse at free or reset, and it will be overwritten when we refill the buffer with something else, but if the user is depending on old outputs not remaining in memory, we might want to cleanse the buffer even before the EVP_MD_CTX is freed or reset.

re 4: Okay, will do.

re 5: I'll give it a shot, though it may take me a bit of time to wrap my head around the test system

levitte commented:
With regards to 3, I imagine that you mean the bytes in `ctx->buf` that have been transferred to the called, if they should be cleansed, possibly a small portion at a time (up to `num` bytes).  This would happen in what is currently called `shake_final`, correct?  Basically, the remainder condition body could look like this:

``` C
    if (remainder) {
        SHA3_squeeze(ctx->A, ctx->buf, bsz, bsz);
        memcpy(md+direct_bytes, ctx->buf, remainder);
        ctx->num = bsz - remainder;
        OPENSSL_cleanse(ctx->buf, ctx->num);
    }
```

Did I catch that correctly?

I'm not sure what to answer...  part of me wants to go "oh bother!", but I'm sure that if we up the paranoia levels (and this is security stuff, where paranoia levels are often set on "high"), it could make sense.  So the question you're asking, I assume, is if it's worth the CPU cycles.

nmathewson commented:
Okay, I've tried to address the issues above per your comments.

re 1: no action.  KeccakF1600 still needs to be exposed from the other assembly files

2. Renamed to EVP_DigestSqueezeXOF; could rename to something else instead if you want.

3. @levitte -- yes, that's what I meant.  I've added a commit to do it, but if you'd rather not take that commit, it's fine.

4. Updated so that once you have squeezed an XOF, you can no longer update it until it is reset.

5. Converted my demo program to a test, and added it to the branch.

levitte commented on crypto/evp/digest.c:
```diff
@@ -179,7 +179,7 @@ int EVP_DigestFinal_ex(EVP_MD_CTX *ctx, unsigned char *md, unsigned int *size)
 }
 
-int EVP_DigestFinalXOF(EVP_MD_CTX *ctx, unsigned char *md, size_t size)
+int EVP_DigestSqueezeXOF(EVP_MD_CTX *ctx, unsigned char *md, size_t size)
```
You should probably check in here that the flag `EVP_MD_CTX_FLAG_CLEANED` isn't already set.

levitte replied:
Say what?  You're setting that flag in the new `EVP_DigestFinalXOF`!

nmathewson replied:
Only when ctx->digest->cleanup is non-NULL.

levitte commented on crypto/evp/m_sha3.c:
```diff
@@ -123,12 +139,88 @@ static int sha3_final(EVP_MD_CTX *evp_ctx, unsigned char *md)
+    /* These bytes will be part of a partial block, copied from buf. */
+    size_t remainder = md_size % bsz;
+    /* We can write this many bytes to md directly. */
+    size_t direct_bytes = md_size - remainder;
```
Sorry, we need to abide to C90 syntax, which means that variable definitions at the beginning of blocks only.

There are two ways to do that.  One is to move the variable definitions (but not the assignments) to the top of the functions, the other is to form an unconditional block that wraps the use of these variables.

levitte commented on test/evp_xof_test.c:
```diff
@@ -0,0 +1,131 @@
+
+    unsigned char buf1[BUFLEN], buf2[BUFLEN];
+
+    // Compute the XOF output in one shot.
```
C90 doesn't have C++ comments, so please change them to C style `/* ... */ ones

levitte commented on test/evp_xof_test.c:
```diff
@@ -0,0 +1,131 @@
+    int i;
+    const int iterations = 512;
+    int result = 1;
+    for (i = 0; i < iterations; ++i) {
```
We prefer one empty line between variable definitions and the rest of the code.

levitte commented on test/evp_xof_test.c:
```diff
@@ -0,0 +1,131 @@
+{
+    int i;
+    int result = 1;
+    for (i = 1; i < 512; ++i) {
```
Here too

levitte commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -111,7 +111,9 @@ EVP_DigestInit_ex() can be called to initialize a new digest operation.
-made, but EVP_DigestInit_ex() can be called to initialize a new operation.
+made, but further calls to EVP_DigestFinalXOF can be made to squeeze more
+output from the function. Further, EVP_DigestInit_ex() can still be
+called to initialize a new operation.
```
Documentation needs to change

levitte commented:
Regarding assembler, it looks like most of them are pretty straight forward:

``` diff
diff --git a/crypto/sha/asm/keccak1600-armv4.pl b/crypto/sha/asm/keccak1600-armv4.pl
index 44504fa8ac..5322dbd322 100755
--- a/crypto/sha/asm/keccak1600-armv4.pl
+++ b/crypto/sha/asm/keccak1600-armv4.pl
@@ -694,6 +694,7 @@ $code.=<<___;
        ldr     pc,[sp,#440]
 .size  KeccakF1600_int,.-KeccakF1600_int
 
+.global        KeccakF1600
 .type  KeccakF1600, %function
 .align 5
 KeccakF1600:
diff --git a/crypto/sha/asm/keccak1600-armv8.pl b/crypto/sha/asm/keccak1600-armv8.pl
index 7baf046843..dfa7e2ee2e 100755
--- a/crypto/sha/asm/keccak1600-armv8.pl
+++ b/crypto/sha/asm/keccak1600-armv8.pl
@@ -295,6 +295,7 @@ $code.=<<___;
        ret
 .size  KeccakF1600_int,.-KeccakF1600_int
 
+.globl KeccakF1600
 .type  KeccakF1600,%function
 .align 5
 KeccakF1600:
diff --git a/crypto/sha/asm/keccak1600-ppc64.pl b/crypto/sha/asm/keccak1600-ppc64.pl
index 5c23841d6a..72891dbd46 100755
--- a/crypto/sha/asm/keccak1600-ppc64.pl
+++ b/crypto/sha/asm/keccak1600-ppc64.pl
@@ -265,6 +265,7 @@ $code.=<<___;
        .byte   0,12,0x14,0,0,0,0,0
 .size  KeccakF1600_int,.-KeccakF1600_int
 
+.globl KeccakF1600
 .type  KeccakF1600,\@function
 .align 5
 KeccakF1600:
diff --git a/crypto/sha/asm/keccak1600-s390x.pl b/crypto/sha/asm/keccak1600-s390x.pl
index 2d5cb417fa..7e109b5d13 100755
--- a/crypto/sha/asm/keccak1600-s390x.pl
+++ b/crypto/sha/asm/keccak1600-s390x.pl
@@ -335,6 +335,7 @@ ___
 }
 {
 $code.=<<___;
+.globl KeccakF1600
 .type  KeccakF1600,\@function
 .align 32
 KeccakF1600:
diff --git a/crypto/sha/asm/keccak1600-x86_64.pl b/crypto/sha/asm/keccak1600-x86_64.pl
index fda4ee1290..dd09c59649 100755
--- a/crypto/sha/asm/keccak1600-x86_64.pl
+++ b/crypto/sha/asm/keccak1600-x86_64.pl
@@ -344,6 +344,7 @@ $code.=<<___;
        ret
 .size  __KeccakF1600,.-__KeccakF1600
 
+.globl KeccakF1600
 .type  KeccakF1600,\@abi-omnipotent
 .align 32
 KeccakF1600:
diff --git a/crypto/sha/asm/keccak1600p8-ppc.pl b/crypto/sha/asm/keccak1600p8-ppc.pl
index 9a97cfdbf8..53d652ecc2 100755
--- a/crypto/sha/asm/keccak1600p8-ppc.pl
+++ b/crypto/sha/asm/keccak1600p8-ppc.pl
@@ -219,6 +219,7 @@ KeccakF1600_int:
        .byte   0,12,0x14,0,0,0,0,0
 .size  KeccakF1600_int,.-KeccakF1600_int
 
+.globl KeccakF1600
 .type  KeccakF1600,\@function
 .align 5
 KeccakF1600:
```

I haven't yet been able to figure out how hard it will be to change the following files, though:

- `crypto/sha/asm/keccak1600-avx2.pl`
- `crypto/sha/asm/keccak1600-avx512.pl`
- `crypto/sha/asm/keccak1600-avx512vl.pl`

kroeckx commented:
I think the other assembler files exist but aren't used.

levitte commented:
> I think the other assembler files exist but aren't used.

Oh you're right!  Ok, ignore for now, then.

nmathewson commented:
I've left a question on the EVP _MD_CTX_FLAG_CLEANED comment, and made the other requested changes.

levitte commented on crypto/evp/digest.c:
```diff
@@ -188,18 +188,28 @@ int EVP_DigestFinalXOF(EVP_MD_CTX *ctx, unsigned char *md, size_t size)
+
+    if (ctx->digest->cleanup != NULL) {
+        ctx->digest->cleanup(ctx);
+        EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_CLEANED);
```
Shouldn't this flag be set regardless, even if `ctx->digest->cleanup == NULL`?

nmathewson replied:
Possibly, but I am pretty sure that I didn't change that.  It's been that way ever since OpenSSL 0.9.7, and I'm not sure I should change it in this ticket.

levitte replied:
Ah ok

nmathewson commented:
I've got Travis to pass, but there's one failing appveyor test. It's not clear to me what's going wrong there, but it does seem to be something to do with the tests.

kroeckx commented:
The log contained:
```
Can't spawn "..\evp_test.exe ..\..\..\test\recipes\30-test_evp_data\evpdigest.txt": No error at C:\projects\openssl\util\perl/OpenSSL/Test.pm line 481.
```

and
```
Can't spawn "..\evp_xof_test.exe": No error at C:\projects\openssl\util\perl/OpenSSL/Test.pm line 481.
```

But the other evp_tests ran without problem, so I assume it's just some appveyor problem and have restarted the test.

I also see this in the log:
```
evp_xof_test.c(69): warning C4267: 'function': conversion from 'size_t' to 'unsigned int', possible loss of data
```
The randint/constint functions should probably take a size_t as parameter, and also return it.

kroeckx commented:
The errors seems to stay, so I assume it's some real problem.

nmathewson commented:
I've fixed the implicit cast warning and added a (non-functional, see above) check for the CLEANED flag. I'm going to look at the appveyor test failure some more, but I might need help from someone who better understands the build and test systems on Windows.

nmathewson commented:
All I've been able to figure out for the bug so far is that it consistently happens with shared x86_64 builds, but not with static builds, and not with shared 32-bit builds.

nmathewson commented:
Based on the symptoms, I tried to guess a fix for the appveyor issue, but my guess was wrong.

I believe that the problem might be related to some windows shared library+assembly issue, since the KeccakF1600 function only exists for windows on 64-bit windows builds, and only shared 64-bit builds are failing.  I tried to fix the problem by changing the abi declaration in the assembly file, but either I did it wrong, or my guess was wrong.

I think I'll need help on this from somebody who knows windows ABI stuff.

Any other remaining issues with this branch?

kroeckx commented:
The static builds don't actually seem to be doing anything, see #7958

nmathewson commented:
Hi. What are our next steps here?

FdaSilvaYY commented on crypto/evp/m_sha3.c:
```diff
@@ -123,12 +139,88 @@ static int sha3_final(EVP_MD_CTX *evp_ctx, unsigned char *md)
+    size_t bsz = ctx->block_size;
+    size_t md_size = ctx->md_size;
+    size_t num = ctx->num;
+    /* We may need to invoke KeccakF directly, since SHA3_squeeze does not
```
Style : comment block like this :
```C
/*
 * We ...
```

FdaSilvaYY commented on test/evp_xof_test.c:
```diff
@@ -0,0 +1,136 @@
+
+#define BUFLEN 2048
+
+/* Test whether the squeeze function of xof is correctly implemented by making
```
style: comment block starts with /* on its own line

FdaSilvaYY commented:
This PR needs a fresh rebase.

kroeckx commented:
Someone needs to figure out what the problem is on Windows.

The non-shared versions are not build by default, but can be build when added "[extended tests]" to the commit message of the last commit. I guess adding that will at least make it more obvious in which cases it all fails. But I guess someone will just have to really look on Windows what the problem is.

nmathewson commented:
@levitte , would you be able to take a look at the remaining Windows issue on this branch?  @richsalz suggested that you might be a good person to ask.

richsalz commented:
@dot-asm any suggestions on making "raw" keccak available in the asm code?

dot-asm commented:
> any suggestions on making "raw" keccak available in the asm code?

In one word, don't. Thing is that even though A[5][5] declaration is available to rest of libcrypto, actual layout is effectively opaque. Even in keccak1600.c, it can be bit-interleaved on 32-bit platforms. In assembly it's even more elaborate, as on top of bit-interleaving, it can be completely different order of elements (e.g. A[1][0] where A[0][1] should be), alternative byte order (e.g. on z14), alternative word order (e.g. on big-endian ARM). And all this can depend on current processor capabilities... In other words, if you want to extend XOF in subsequent call[s], then what is suggested here is not the way to do it. A[5][5] should be viewed as opaque, and one should let any given module to maintain consistency internally. It would be more appropriate to specify how interface at keccak1600.c level should be extended to permit this. For example one can say that SHA3_squeeze should return a value, which could be then passed back to SHA3_squeeze, so that self-consistency can be maintained internally. It would be somewhat tricky to cover most general case... As simpler alternative, one can say that it would be sufficient if SHA3_squeeze accepted an extra boolean, that would tell it to skip initial loop, go straight to lowest-level block call and only then go to the first loop. In this case it would be caller's responsibility(*) to pull exactly `n*block_size` bytes in each call, naturally with one boolean value in initial call, and alternative one in subsequent ones. With this in mind it should also be noted that it's actually possible to have SHA3_squeeze extend XOF even as it stands now. All one needs to do is to ignore first block_size bytes in subsequent calls. I mean 2nd (3rd, etc.) would have to ask for `(n+1)*block_size` and ignore first block.

(*) "Caller's responsibility" doesn't have to mean "**application** responsibility" as one can, and really should, resolve this in EVP. I mean you can let application go wild and make it EVP's responsibility to make right calls.

baentsch commented:
May I ask what it would take to "reactivate" this PR (or the related issue #7894))?

Background: @beldmit noted the performance deficiencies when using OpenSSL as the SHA3 crypto engine for running PQ crypto (via `liboqs`: https://github.com/open-quantum-safe/liboqs/issues/1426).

So far, we worked around this with our own (thanks to @jschanck) XKCP integration. But in the interest of aligning implementations, particularly when using PQC in an OpenSSL provider, it would make sense to save code & effort and improve usability/performance of the "OpenSSL built-in" SHA3/Shake.

Initial question: Can/should this be done with a new API or in another way? 

Thanks in advance for any suggestions whether & how to move this forward, @levitte @paulidale @t8m

paulidale commented:
@beldmit mentioned this before the OTC call yesterday and the consensus was that we should do this.

I'm not sure how to handle the various assembly implementations.  It's not going to be easy -- see @dot-asm's comment above.  Personally, I don't have an issue with not all platforms supporting a true extensible digest mode -- i.e. those platforms where the assembly can't easily be adapted, don't get the support.

One question/thought: would a completely separate XOF API make sense?  I'll note that I don't object to a reentrant DigestFinalXOF for XOF digests, I can't see any problems.  However, AEAD encryption was shoehorned into the existing symmetric cipher APIs and (IMO at least) it would have been better done as a completely separate new API.  I'd like to make sure we're not repeating a similar mistake here.

baentsch commented:
> the consensus was that we should do this.

Thanks for letting me know.

> i.e. those platforms where the assembly can't easily be adapted, don't get the support.

Absolutely agreed. That's precisely how we do it in `liboqs`: If there's no interest to do optimizations on/for a particular platform, we're running "generic C code". It can take years (or a student's master thesis :-) to "upgrade".

> would a completely separate XOF API make sense? 

I for one wouldn't complain -- particularly if it were close to what we currently use :-) See https://github.com/open-quantum-safe/liboqs/blob/main/src/common/sha3/sha3.h

t8m commented:
IMO the decision point on whether a new front-end API call or just new provider API call should be added is the question how easy it would be to make the subsequent DigestFinalXOF() calls return error on implementations that do not support the subsequent calls properly. IMO from an application perspective it would be more logical to keep using the same call for the first and subsequent squeezes, but perhaps it would be a good idea to add a separate provider API call for the further squeezes. I.e., have a flag in EVP_MD_CTX whether the digest is finalized and if not, call the existing OSSL_FUNC_digest_final_fn() but for the subsequent calls call a new provider API function OSSL_FUNC_digest_squeeze_fn() which would not exist for implementations that do not provide this functionality and so the call would naturally fail.

kroeckx commented:
#13470 seems to be related to this.

slontis commented:
Putting a OTC hold on this so it is discussed in an OTC meeting. How do we get this PR moving again?

nmathewson commented:
(FWIW, this is not something I remember well enough to work on it any longer. Anybody else should feel free to reuse as much or as little of this branch as seems useful, but I don't think I'll have the time to pick it up again any time soon.)

t8m commented:
**OTC: We've discussed this and @slontis will create some proposal how to move forward.**

t8m commented:
#13470 is related

slontis commented:
I have attempted to create a document to describe proposals. See PR #21443

slontis commented:
@nmathewson Thankyou for writing this PR. I will be sure to mention you in any PR that is derived from this work.

t8m commented:
We have streaming SHAKE now.
